cpp/src/arrow/engine/expression_test.cc

```diff
@@ -0,0 +1,265 @@
+
+  auto table = ExprType::Table(s);
+  EXPECT_EQ(table.shape(), ExprType::Shape::TABLE);
+  EXPECT_TRUE(table.schema()->Equals(s));
+  EXPECT_FALSE(table.IsScalar());
+  EXPECT_FALSE(table.IsArray());
+  EXPECT_TRUE(table.IsTable());
+}
+
+TEST_F(ExprTypeTest, IsPredicate) {
+  auto bool_scalar = ExprType::Scalar(boolean());
+  EXPECT_TRUE(bool_scalar.IsPredicate());
+
+  auto bool_array = ExprType::Array(boolean());
+  EXPECT_TRUE(bool_array.IsPredicate());
+
+  auto bool_table = ExprType::Table(schema({field("b", boolean())}));
+  EXPECT_FALSE(bool_table.IsPredicate());
+
+  auto i32_scalar = ExprType::Scalar(int32());
+  EXPECT_FALSE(i32_scalar.IsPredicate());
+}
+
+TEST_F(ExprTypeTest, Broadcast) {
+  auto bool_scalar = ExprType::Scalar(boolean());
+  auto bool_array = ExprType::Array(boolean());
+  auto bool_table = ExprType::Table(schema({field("b", boolean())}));
+  auto i32_scalar = ExprType::Scalar(int32());
+
+  EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid, HasSubstr("operands must be of same type"),
+                                  ExprType::Broadcast(bool_scalar, i32_scalar));
+  EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid, HasSubstr("operands must not be tables"),
+                                  ExprType::Broadcast(bool_scalar, bool_table));
+
+  EXPECT_THAT(ExprType::Broadcast(bool_scalar, bool_scalar), OkAndEq(bool_scalar));
+  EXPECT_THAT(ExprType::Broadcast(bool_scalar, bool_array), OkAndEq(bool_array));
+  EXPECT_THAT(ExprType::Broadcast(bool_array, bool_scalar), OkAndEq(bool_array));
+  EXPECT_THAT(ExprType::Broadcast(bool_array, bool_array), OkAndEq(bool_array));
+}
+
+TEST_F(ExprTypeTest, WithTypeOrSchema) {
+  auto bool_scalar = ExprType::Scalar(boolean());
+  auto bool_array = ExprType::Array(boolean());
+  auto bool_table = ExprType::Table(schema({field("b", boolean())}));
+
+  auto i32 = int32();
+  auto other = schema({field("a", i32)});
+
+  EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid, HasSubstr("Cannot cast a ScalarType with"),
+                                  bool_scalar.WithSchema(other));
+  EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid, HasSubstr("Cannot cast an ArrayType with"),
+                                  bool_array.WithSchema(other));
+  EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid, HasSubstr("Cannot cast a TableType with"),
+                                  bool_table.WithType(i32));
+
+  EXPECT_EQ(bool_scalar.WithType(i32), ExprType::Scalar(i32));
+  EXPECT_EQ(bool_array.WithType(i32), ExprType::Array(i32));
+  EXPECT_EQ(bool_table.WithSchema(other), ExprType::Table(other));
+}
+
+class ExprTest : public testing::Test {};
+
+TEST_F(ExprTest, ScalarExpr) {
+  ASSERT_RAISES(Invalid, ScalarExpr::Make(nullptr));
+
+  auto i32 = int32();
+  ASSERT_OK_AND_ASSIGN(auto value, MakeScalar(i32, 10));
+  ASSERT_OK_AND_ASSIGN(auto expr, ScalarExpr::Make(value));
+  EXPECT_EQ(expr->kind(), ExprKind::SCALAR_LITERAL);
+  EXPECT_EQ(expr->type(), ExprType::Scalar(i32));
+  EXPECT_EQ(*expr->scalar(), *value);
+}
```

Comment:
As I stated in the header using `ScalarExpr` for a literal is misleading -- need to think about scalar values resulting column expressions, e.g.

https://cloud.google.com/bigquery/docs/reference/standard-sql/arrays

BigQuery lets you do

```
SELECT some_numbers,
  (SELECT SUM(x)
   FROM UNNEST(s.some_numbers) x) AS sums
FROM sequences s;
```

It might be nice to be able to do (pseudocode)

```
GroupBy(
  Projection(t, {FieldRef("key"), ArraySum(FieldRef("a")) -> ScalarExpr<Int64>}),
  /*group_exprs=*/{FieldRef("key")},
  /*agg_exprs=*/{Sum(FieldRef("a"))}
)
```

Or if "a" is a struct like

```
a: struct<f0: list<int64>, f1: string>
```

then we want to be able to do

```
ArraySum(
  // Select a column path
  FieldRef({"a", "f0"}) -> list<int64>
) -> int64
```